thread.c

```diff
@@ -630,14 +635,70 @@ static void thread_libevent_process(evutil_socket_t fd, short which, void *arg)
     }
 }
 
-// Interface is slightly different on various platforms.
-// On linux, at least, the len limit is 16 bytes.
 #define THR_NAME_MAXLEN 16
-void thread_setname(pthread_t thread, const char *name) {
-assert(strlen(name) < THR_NAME_MAXLEN);
+static void thread_setname(pthread_t thread, const char *name) {
+  assert(strlen(name) < THR_NAME_MAXLEN);
+
 #if defined(__linux__)
-pthread_setname_np(thread, name);
+#ifdef HAVE_PTHREAD_SETNAME_NP
+  pthread_setname_np(thread, name);
+#else
+  // If pthread_setname_np is not available, use prctl to set the thread name.
+  // Note that this can only be called by the thread itself.
+  if (pthread_self() == thread) {
+    prctl(PR_SET_NAME, (unsigned long)name, 0, 0, 0);
+  } else {
+    fprintf(stderr, "Threads other than the caller cannot be renamed [%s].", name);
+  }
 #endif
+#elif defined(__FreeBSD__)
+  pthread_set_name_np(thread, name);
+#elif defined(__APPLE__) && defined(__MACH__)
+  pthread_setname_np(name);
```

Comment:
does it work as expected on macOS ?

Reply:
Yes.
On MACOS, the function pthread_setname_np(const char*) sets the internal name of the calling thread to the string value specified by the name argument.
Type is done with the prctl.

Now, if the rest of the source compiled under MACOS, that's another story.

Reply:
> Yes. On MACOS, the function pthread_setname_np(const char*) sets the internal name of the calling thread
 Sure https://github.com/apple/darwin-libpthread/blob/main/src/pthread.c#L1140 but what I meant is Linux/FreeBSD allows to specify a specific thread rather than the calling one.
> 
> Now, if the rest of the source compiled under MACOS, that's another story.

🙂

Reply:
It should build on recent MacOS: if it doesn't please report a bug.

Reply:
> Sure https://github.com/apple/darwin-libpthread/blob/main/src/pthread.c#L1140 but what I meant is Linux/FreeBSD allows to specify a specific thread rather than the calling one.

Yes, but as long as you have GLIBC >= 2.12. Por isso o tratamento acima usando prctl ou pthread_setname_np on Linux.

Reply:
fair point even tough glibc 2.12 is already fairly old.